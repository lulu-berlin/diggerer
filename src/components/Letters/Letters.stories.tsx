import React from "react";
import { storiesOf } from "@storybook/react";
import { LettersSymbols } from "components/Letters/LettersSymbols";
import { Letters } from "components/Letters/Letters";

console.log(module);

storiesOf("Letters", module)
  .add("hello world", () => (
    <>
      <LettersSymbols />
      <Letters charWidth={16} charHeight={16}>
        hello world
      </Letters>
    </>
  ))
  .add("DIGGER", () => (
    <>
      <LettersSymbols />
      <Letters charWidth={16} charHeight={16}>
        DIGGER
      </Letters>
    </>
  ))
  .add("HIGH SCORE", () => (
    <>
      <LettersSymbols />
      <Letters charWidth={16} charHeight={16}>
        HIGH SCORE
      </Letters>
    </>
  ));
