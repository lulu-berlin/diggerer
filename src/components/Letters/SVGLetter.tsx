import React from "react";

export type LetterProps = React.SVGProps<SVGUseElement> & {
  /**
   * The letter to render (a-z or A-Z).
   */
  char: string;
};

/**
 * This component renders a letter from the SVG symbols created by <LettersSymbols>.
 */
export const SVGLetter: React.FC<LetterProps> = ({ char, ...svgProps }) => {
  const letter = char.toUpperCase();

  return /[A-Z]/.test(letter) ? (
    <use href={`#CLET${char.toUpperCase()}`} {...svgProps} />
  ) : null;
};
