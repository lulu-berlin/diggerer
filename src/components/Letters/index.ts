export * from "./LettersSymbols";
export * from "./Letters";
