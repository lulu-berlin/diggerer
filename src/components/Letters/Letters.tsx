import React from "react";
import { SVGLetter } from "components/Letters/SVGLetter";

export type LettersProps = {
  children: string;
  charWidth: number;
  charHeight: number;
};

export const Letters: React.FC<LettersProps> = ({
  children: text,
  charWidth,
  charHeight
}) => (
  <svg height={charHeight} width={charWidth * text.length}>
    {text.split("").map((char, i) => (
      <SVGLetter
        char={char}
        x={charWidth * i}
        y={0}
        width={charWidth}
        height={charHeight}
        key={i}
      />
    ))}
  </svg>
);
