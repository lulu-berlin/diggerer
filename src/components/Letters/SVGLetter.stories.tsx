import React from "react";
import { storiesOf } from "@storybook/react";
import { LettersSymbols } from "components/Letters/LettersSymbols";
import { SVGLetter } from "components/Letters/SVGLetter";

const A = "A".charCodeAt(0);
const Z = "Z".charCodeAt(0);

Array.from({ length: Z - A + 1 }).forEach((_, i) => {
  const char = String.fromCharCode(A + i);

  storiesOf("SVGLetter", module).add(char, () => (
    <>
      <LettersSymbols />
      <svg width={16} height={16}>
        <SVGLetter x={0} y={0} width={16} height={16} char={char} />
      </svg>
    </>
  ));
});
