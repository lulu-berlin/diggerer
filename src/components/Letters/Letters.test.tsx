import React from "react";
import { shallow } from "enzyme";
import { Letters } from "components/Letters";
import { SVGLetter } from "components/Letters/SVGLetter";

it('should render "hello world" correctly', () => {
  const helloWorld = "hello world";
  const charWidth = 10;
  const charHeight = 20;
  const wrapper = shallow(
    <Letters charWidth={charWidth} charHeight={charHeight}>
      {helloWorld}
    </Letters>
  );

  expect(wrapper).toMatchSnapshot();

  const svg = wrapper.find("svg").at(0);
  expect(svg).toHaveLength(1);

  expect(svg.props()).toMatchObject({
    width: charWidth * helloWorld.length,
    height: charHeight
  });

  const letters = wrapper.find(SVGLetter);

  expect(letters).toHaveLength(helloWorld.length);

  for (let i = 0; i < helloWorld.length; ++i) {
    expect(letters.at(i).props()).toStrictEqual({
      char: helloWorld[i],
      height: charHeight,
      width: charWidth,
      x: i * charWidth,
      y: 0
    });
  }
});
