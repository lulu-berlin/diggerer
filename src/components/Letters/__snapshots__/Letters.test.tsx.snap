// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render "hello world" correctly 1`] = `
<svg
  height={20}
  width={110}
>
  <SVGLetter
    char="h"
    height={20}
    key="0"
    width={10}
    x={0}
    y={0}
  />
  <SVGLetter
    char="e"
    height={20}
    key="1"
    width={10}
    x={10}
    y={0}
  />
  <SVGLetter
    char="l"
    height={20}
    key="2"
    width={10}
    x={20}
    y={0}
  />
  <SVGLetter
    char="l"
    height={20}
    key="3"
    width={10}
    x={30}
    y={0}
  />
  <SVGLetter
    char="o"
    height={20}
    key="4"
    width={10}
    x={40}
    y={0}
  />
  <SVGLetter
    char=" "
    height={20}
    key="5"
    width={10}
    x={50}
    y={0}
  />
  <SVGLetter
    char="w"
    height={20}
    key="6"
    width={10}
    x={60}
    y={0}
  />
  <SVGLetter
    char="o"
    height={20}
    key="7"
    width={10}
    x={70}
    y={0}
  />
  <SVGLetter
    char="r"
    height={20}
    key="8"
    width={10}
    x={80}
    y={0}
  />
  <SVGLetter
    char="l"
    height={20}
    key="9"
    width={10}
    x={90}
    y={0}
  />
  <SVGLetter
    char="d"
    height={20}
    key="10"
    width={10}
    x={100}
    y={0}
  />
</svg>
`;
